lib/vector-core/src/schema/requirement.rs

```diff
@@ -32,8 +52,316 @@ impl Requirement {
+            (
+                "missing required meaning",
+                TestCase {
+                    requirement: Requirement::empty().required_meaning("foo", Kind::any()),
+                    definition: Definition::empty(),
+                    errors: vec![ValidationError::MeaningMissing { identifier: "foo" }],
+                },
+            ),
+            (
+                "missing required meanings",
+                TestCase {
+                    requirement: Requirement::empty()
+                        .required_meaning("foo", Kind::any())
+                        .required_meaning("bar", Kind::any()),
+                    definition: Definition::empty(),
+                    errors: vec![
+                        ValidationError::MeaningMissing { identifier: "bar" },
+                        ValidationError::MeaningMissing { identifier: "foo" },
+                    ],
+                },
+            ),
+            (
+                "missing optional meaning",
+                TestCase {
+                    requirement: Requirement::empty().optional_meaning("foo", Kind::any()),
+                    definition: Definition::empty(),
+                    errors: vec![],
+                },
+            ),
+            (
+                "missing mixed meanings",
+                TestCase {
+                    requirement: Requirement::empty()
+                        .optional_meaning("foo", Kind::any())
+                        .required_meaning("bar", Kind::any()),
+                    definition: Definition::empty(),
+                    errors: vec![ValidationError::MeaningMissing { identifier: "bar" }],
+                },
+            ),
+            (
+                "invalid required meaning kind",
+                TestCase {
+                    requirement: Requirement::empty().required_meaning("foo", Kind::boolean()),
+                    definition: Definition::empty().required_field(
+                        "foo",
+                        Kind::integer(),
+                        Some("foo"),
+                    ),
+                    errors: vec![ValidationError::MeaningKind {
+                        identifier: "foo",
+                        want: Kind::boolean(),
+                        got: Kind::integer(),
+                    }],
+                },
+            ),
+            (
+                "invalid optional meaning kind",
+                TestCase {
+                    requirement: Requirement::empty().optional_meaning("foo", Kind::boolean()),
+                    definition: Definition::empty().required_field(
+                        "foo",
+                        Kind::integer(),
+                        Some("foo"),
+                    ),
+                    errors: vec![ValidationError::MeaningKind {
+                        identifier: "foo",
+                        want: Kind::boolean(),
+                        got: Kind::integer(),
+                    }],
+                },
```

Comment:
I'm uncertain if I quite follow what your thoughts are here.

Essentially, this is the same as other type checks, but in this case, the sink defines the field as optional, but since the definition has the field set as present, but as the wrong type, it's an error.

What gave you pause in questioning whether this should be an error?

Reply:
It's the former, see also the description of the `optional_meaning` method.